File: Diagrammatize.Design.Pattern/code/cxx/Builder/A3/MarkdownBuilder.cpp
#include "MarkdownBuilder.hpp"
#include <iostream>

MarkdownBuilder::~MarkdownBuilder() {
}

void MarkdownBuilder::makeTitle(const std::string& title) {                 // Markdown文件的标题
    filename = title + ".md";                                                 // 将标题作为文件名
    writer.exceptions(std::ifstream::failbit | std::ifstream::badbit);
    try {
        writer.open(filename);                                                      // 打开 ofstream
    } catch (const std::ios_base::failure& e) {
        std::cerr << "open " << filename << " failed: " << e.what() << "\n";
        throw;
    }
    writer << "- - -" << '\n';          // 分割线
    writer << "# " << title << '\n';    // 输出标题
    writer << '\n';
}

void MarkdownBuilder::makeString(const std::string& str) {                  // Markdown文件中的字符串
    writer << "> " << str << '\n';                                   // 用<p>标签输出
    writer << '\n';
}

void MarkdownBuilder::makeItems(const std::vector<std::string>& items) {    // Markdown文件中的条目
    for (int i = 0; i < items.size(); i++) {
        writer << "* " << items[i] << '\n';
    }
    writer << '\n';
}

void MarkdownBuilder::close() {                                             // 完成文档
    writer << "- - -" << '\n';                                                  // 分割线
    writer.close();                                                             // 关闭文件
}

std::string MarkdownBuilder::getResult() {                                  // 编写完成的文档
    return filename;                                                            // 返回文件名
}

File: Diagrammatize.Design.Pattern/code/python/Builder/A3/MarkdownBuilder.py
import traceback

class MarkdownBuilder:
    def __init__(self):
        self.filename = ''          # 文件名
        self.writer = None          # 用于编写文件的writer

    def makeTitle(self, title):         # Markdown文件的标题
        self.filename = title + ".md"     # 将标题作为文件名
        try: 
            self.writer = open(self.filename, 'w')    # 生成 writer
        except:
            traceback.print_exc()
            raise
            
        print("- - -", file=self.writer)            # 分割线
        print("# " + title, file=self.writer)       # 输出标题
        print("", file=self.writer)

    def makeString(self, str_):                     # Markdown文件中的字符串
        print("> " + str_, file=self.writer)
        print("", file=self.writer)

    def makeItems(self, items):                     # Markdown文件中的条目
        for i in range(len(items)):
            print("* " + items[i], file=self.writer)
        print("", file=self.writer)

    def close(self):                        # 完成文档
        print("- - -", file=self.writer)        # 分割线
        self.writer.close()                     # 关闭文件

    def getResult(self):                    # 编写完成的文档
        return self.filename                    # 返回文件名
